include/dali/core/float16.h

```diff
@@ -17,31 +17,313 @@
 
 #include <cuda_fp16.h>  // for __half & related methods
 #include <type_traits>
+#include "dali/core/host_dev.h"
+#include "dali/core/force_inline.h"
 #ifndef __CUDA_ARCH__
 #include "dali/util/half.hpp"
 #endif
 
 namespace dali {
 
-// For the GPU
+struct float16 {
+  DALI_HOST_DEV
+  float16() = default;
+  float16(const float16 &) = default;
+
+#ifdef __CUDA_ARCH__
+  __device__ float16(__half x)          : impl(x) {}  // NOLINT
+  __device__ float16(float x)           : impl(x) {}  // NOLINT
+  __device__ float16(double x)          : impl(x) {}  // NOLINT
+  __device__ float16(signed char x)     : impl(x) {}  // NOLINT
+  __device__ float16(unsigned char x)   : impl(x) {}  // NOLINT
+  __device__ float16(short x)           : impl(x) {}  // NOLINT
+  __device__ float16(unsigned short x)  : impl(x) {}  // NOLINT
+  __device__ float16(int x)             : impl(x) {}  // NOLINT
+  __device__ float16(unsigned int x)    : impl(x) {}  // NOLINT
+  __device__ float16(long x)            : impl(static_cast<long long>(x)) {}  // NOLINT
+  __device__ float16(unsigned long x)   : impl(static_cast<unsigned long long>(x)) {}  // NOLINT
+  __device__ float16(long long x)           : impl(x) {}  // NOLINT
+  __device__ float16(unsigned long long x)  : impl(x) {}  // NOLINT
+#endif
+#ifndef __CUDA_ARCH__
+  __host__ float16(half_float::half x) : impl(x) {}  // NOLINT
+  __host__ float16(float x)           : impl(x) {}  // NOLINT
+  __host__ float16(double x)          : impl(x) {}  // NOLINT
+  __host__ float16(signed char x)     : impl(x) {}  // NOLINT
+  __host__ float16(unsigned char x)   : impl(x) {}  // NOLINT
+  __host__ float16(short x)           : impl(x) {}  // NOLINT
+  __host__ float16(unsigned short x)  : impl(x) {}  // NOLINT
+  __host__ float16(int x)             : impl(x) {}  // NOLINT
+  __host__ float16(unsigned int x)    : impl(x) {}  // NOLINT
+  __host__ float16(long x)            : impl(static_cast<int64_t>(x)) {}  // NOLINT
+  __host__ float16(unsigned long x)   : impl(static_cast<uint64_t>(x)) {}  // NOLINT
+  __host__ float16(long long x)          : impl(static_cast<int64_t>(x)) {}  // NOLINT
+  __host__ float16(unsigned long long x) : impl(static_cast<uint64_t>(x)) {}  // NOLINT
+#endif
+
+
+#ifdef __CUDA_ARCH__
+  using impl_t = __half;
+#else
+  using impl_t = half_float::half;
+#endif
+
+  impl_t impl;
+
+  DALI_HOST_DEV DALI_FORCEINLINE
+  constexpr operator impl_t() const noexcept { return impl; }
+
+#ifdef __CUDA_ARCH__
+  DALI_FORCEINLINE
+  __device__ operator float() const noexcept { return impl; }
```

Comment:
Does it run the desired conversion when the `float` is constructed from the `impl`? 

Just checking.